packages/react/src/use-chat.ts

```diff
@@ -190,38 +201,78 @@ Default is undefined, which disables throttling.
     [stableInitialMessages],
   );
 
-  // Store the chat state in SWR, using the chatId as the key to share states.
-  const { data: messages, mutate } = useSWR<UIMessage[]>(
-    [chatKey, 'messages'],
-    null,
-    { fallbackData: processedInitialMessages },
-  );
+  const newStoreRef = useRef<ChatStore>();
+  const chatStore = useMemo(() => {
+    if (store) {
+      if (!store.hasChat(chatId)) {
+        store.addChat(chatId, processedInitialMessages);
+      }
+      return store;
+    }
 
-  // Keep the latest messages in a ref.
-  const messagesRef = useRef<UIMessage[]>(messages || []);
-  useEffect(() => {
-    messagesRef.current = messages || [];
-  }, [messages]);
+    if (!newStoreRef.current) {
+      newStoreRef.current = new ChatStore({
+        chats: { [chatId]: { messages: processedInitialMessages } },
+        getCurrentDate,
+      });
+    }
+
+    return newStoreRef.current;
+  }, [store, chatId, processedInitialMessages, getCurrentDate]);
 
-  // stream data
+  const subscribeToMessages = useCallback(
+    (cb: () => void) => {
+      return chatStore.subscribe({
+        id: chatId,
+        onChatChanged: () => cb(),
+      });
+    },
+    [chatStore, chatId],
+  );
+  const messages = useSyncExternalStore(
+    subscribeToMessages,
+    () => chatStore.getMessages(chatId) ?? EMPTY_MESSAGES,
+    () => processedInitialMessages,
+  );
+  const status: 'submitted' | 'streaming' | 'ready' | 'error' =
+    useSyncExternalStore(
+      cb =>
+        chatStore.subscribe({
+          id: chatId,
+          onChatChanged: e => {
+            if (e === ChatStoreEvent.ChatStatusChanged) {
+              cb();
+            }
+          },
+        }),
+      () => chatStore.getStatus(chatId) ?? 'ready',
+      () => 'ready',
+    );
+  const error = useSyncExternalStore(
+    cb =>
+      chatStore.subscribe({
+        id: chatId,
```

Comment:
how would id changes work? (switching the useChat hook to a different id)

Reply:
good flag! would love your take on the hook approach to chat store i took (used inside useChat)